### PR TITLE
add user upsertable parameter to data sources

### PR DESCRIPTION
**Pull request by Uzay-G**
For use in creating data sources users can add to, for example in the extension.

### CODE REVIEW

**Comment by Uzay-G** [comment]
image

Would appreciate comments on how this should look?

File: front/pages/w/[wId]/ds/new.tsx [hunk 1]
```diff
@@ -292,6 +294,33 @@ export default function DataSourceNew({
...
+                        type="checkbox"
+                        className="h-4 w-4 cursor-pointer border-gray-300 text-violet-600 focus:ring-violet-500"
+                        checked={userUpsertable}
+                        onChange={(e) => setUserUpsertable(e.target.checked)}
+                      />
+                      <label
+                        htmlFor="upsertable"
+                        className="ml-3 block text-sm font-medium text-gray-700"
+                      >
+                        User-upsertable
```
Comment (spolu) [inline front/pages/w/[wId]/ds/new.tsx:315]: Let's call it "Users can upsert"

**Review by spolu — CHANGES_REQUESTED** [review]
Nice job

UI:
- Let's move "User upsertable" checkbox below the other configs (same section, no horizontal bar)
- Let's call it "Users can upsert" in the UI, the explanation text is fine :+1: 

Code:
- Have you checked that running init_db does populate correctly the `false` default value to pre-existing data sources?
- I think we want to let non builder users be able to navigate directly to the data source and update it from the "Builder" interface. Probably a question of adding a check for "userUpsertable" at a few more places? Unless that's already the case?

File: front/pages/w/[wId]/ds/new.tsx (hunk 1 above)
Comment (Uzay-G) [reply front/pages/w/[wId]/ds/new.tsx:315]: Maybe we should say upload instead of upsert? less technical

Comment (spolu) [reply front/pages/w/[wId]/ds/new.tsx:315]: Definitely :+1:

**Comment by Uzay-G** [comment]
Yeah I checked the DB. Will look at your second point.

**Comment by Uzay-G** [comment]
What should the user see then? The builder interface with just upsertable data sources, and then making sure that the user can only upload?

**Comment by Uzay-G** [comment]
Ok I propagated those changes

File: front/pages/w/[wId]/index.tsx [hunk 2]
```diff
@@ -17,17 +18,29 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
     };
   }
 
-  if (owner.role === "user") {
+  // check if upsertable data sources using getDataSources
```
Comment (spolu) [inline front/pages/w/[wId]/index.tsx:21]: Let's not change the default redirect. We still want to go to /u for users. If they have data sources and want to update them the builders/admin of their team will share the URL with them for now :+1:

File: front/types/data_source.ts [hunk 3]
```diff
@@ -10,4 +10,5 @@ export type DataSourceType = {
   dustAPIProjectId: string;
   connectorId?: string;
   connectorProvider?: ConnectorProvider;
+  userUpsertable?: boolean;
```
Comment (spolu) [inline front/types/data_source.ts:13]: Should be `userUpsertable` since the boolean is required in DB

File: front/lib/models.ts [hunk 4]
```diff
@@ -539,8 +539,8 @@ export class DataSource extends Model<
   declare dustAPIProjectId: string;
   declare connectorId?: string;
   declare connectorProvider?: ConnectorProvider;
-
   declare workspaceId: ForeignKey<Workspace["id"]>;
+  declare userUpsertable: CreationOptional<boolean>;
```
Comment (spolu) [inline front/lib/models.ts:543]: Let;s move it before the foreignkey (one line above)

**Review by spolu — CHANGES_REQUESTED** [review]
A few final changes that I didn't catch before. Overall looks good otherwise.

File: front/lib/models.ts [hunk 5]
```diff
@@ -539,7 +539,7 @@ export class DataSource extends Model<
   declare dustAPIProjectId: string;
   declare connectorId?: string;
   declare connectorProvider?: ConnectorProvider;
-
+  declare userUpsertable: CreationOptional<boolean>;
```
Comment (spolu) [inline front/lib/models.ts:542]: We'll want to revert this to `boolean` once it's deployed :+1:

File: front/pages/w/[wId]/ds/new.tsx [hunk 6]
```diff
@@ -347,6 +349,31 @@ export default function DataSourceNew({
                       </p>
                     </div>
                   </div>
+                  <fieldset className="mt-6">
```
Comment (spolu) [inline front/pages/w/[wId]/ds/new.tsx:352]: Can you share an updated screenshot.

I think we don't need the "Upload rights" section thingy

File: front/pages/w/[wId]/index.tsx [hunk 7]
```diff
@@ -20,14 +20,14 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
   if (owner.role === "user") {
     return {
       redirect: {
-        destination: `/w/${context.query.wId}/u`,
+        destination: `/w/${context.query.wId}/ds`,
```
Comment (spolu) [inline front/pages/w/[wId]/index.tsx:23]: nope we want /u here

File: front/pages/w/[wId]/index.tsx [hunk 8]
```diff
@@ -20,14 +20,14 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
...
         permanent: false,
       },
     };
   }
   return {
     redirect: {
-      destination: `/w/${context.query.wId}/a`,
+      destination: `/w/${context.query.wId}/u`,
```
Comment (spolu) [inline front/pages/w/[wId]/index.tsx:30]: Nope we want /a here

**Review by spolu — CHANGES_REQUESTED** [review]
Left a few change requests

File: front/pages/w/[wId]/ds/new.tsx (hunk 6 above)
Comment (Uzay-G) [reply front/pages/w/[wId]/ds/new.tsx:352]: image

vs

image

**Review by spolu — APPROVED** [review]
LGTM

**Comment by spolu** [comment]
Per IRL To be merged tomorrow morning to run through the deploy procedure together

**Comment by spolu** [comment]
Deploying now:
- run front/admin/init_db.sh on prod instances
- deploy front

**Pull request merged**